Unable to insert into transactional table without partitions

I am hitting an error trying to insert values into a transactional table without partitions.

Here's the SQL:

```sql
CREATE TABLE t (x INT) WITH (transactional = true);
INSERT INTO t VALUES (1), (2);
```

This is error message I get:

```
trino:default> CREATE TABLE t (x INT) WITH (transactional = true);
CREATE TABLE
trino:default> INSERT INTO t VALUES (1), (2);

Query 20231015_150350_01245_4bjpf, FAILED, 1 node
Splits: 36 total, 35 done (97.22%)
0.09 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20231015_150350_01245_4bjpf failed: Exception writing _orc_acid_version file for delta directory: s3://warehouse/t//delta_0000001_0000001_0000
```

Looking at `docker compose logs -f trino` shows this:

<details>
<summary>Trino logs</summary>

```
ibis-trino-1  | io.trino.spi.TrinoException: Exception writing _orc_acid_version file for delta directory: s3://warehouse/t//delta_0000001_0000001_0000
ibis-trino-1  |         at io.trino.plugin.hive.HiveMetadata.createOrcAcidVersionFile(HiveMetadata.java:2329)
ibis-trino-1  |         at io.trino.plugin.hive.HiveMetadata.finishInsert(HiveMetadata.java:2135)
ibis-trino-1  |         at io.trino.plugin.base.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:590)
ibis-trino-1  |         at io.trino.tracing.TracingConnectorMetadata.finishInsert(TracingConnectorMetadata.java:668)
ibis-trino-1  |         at io.trino.metadata.MetadataManager.finishInsert(MetadataManager.java:1087)
ibis-trino-1  |         at io.trino.tracing.TracingMetadata.finishInsert(TracingMetadata.java:683)
ibis-trino-1  |         at io.trino.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$4(LocalExecutionPlanner.java:4196)
ibis-trino-1  |         at io.trino.operator.TableFinishOperator.getOutput(TableFinishOperator.java:319)
ibis-trino-1  |         at io.trino.operator.Driver.processInternal(Driver.java:395)
ibis-trino-1  |         at io.trino.operator.Driver.lambda$process$8(Driver.java:298)
ibis-trino-1  |         at io.trino.operator.Driver.tryWithLock(Driver.java:694)
ibis-trino-1  |         at io.trino.operator.Driver.process(Driver.java:290)
ibis-trino-1  |         at io.trino.operator.Driver.processForDuration(Driver.java:261)
ibis-trino-1  |         at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
ibis-trino-1  |         at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
ibis-trino-1  |         at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
ibis-trino-1  |         at io.trino.$gen.Trino_429____20231015_145128_2.run(Unknown Source)
ibis-trino-1  |         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
ibis-trino-1  |         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
ibis-trino-1  |         at java.base/java.lang.Thread.run(Thread.java:833)
ibis-trino-1  | Caused by: java.io.IOException: com.amazonaws.services.s3.model.AmazonS3Exception: Object name contains unsupported characters. (Service: Amazon S3; Status Code: 400; Error Code: XMinioInvalidObjectName; Request ID: 178E502539DD4537; S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8; Proxy: null), S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8
ibis-trino-1  |         at io.trino.hdfs.s3.TrinoS3FileSystem$TrinoS3StreamingOutputStream.flushBuffer(TrinoS3FileSystem.java:1914)
ibis-trino-1  |         at io.trino.hdfs.s3.TrinoS3FileSystem$TrinoS3StreamingOutputStream.close(TrinoS3FileSystem.java:1859)
ibis-trino-1  |         at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.close(FSDataOutputStream.java:77)
ibis-trino-1  |         at org.apache.hadoop.fs.FSDataOutputStream.close(FSDataOutputStream.java:106)
ibis-trino-1  |         at org.apache.hadoop.fs.FSDataOutputStream$PositionCache.close(FSDataOutputStream.java:77)
ibis-trino-1  |         at org.apache.hadoop.fs.FSDataOutputStream.close(FSDataOutputStream.java:106)
ibis-trino-1  |         at io.trino.filesystem.hdfs.HdfsOutputStream.close(HdfsOutputStream.java:87)
ibis-trino-1  |         at io.trino.plugin.hive.util.AcidTables.writeAcidVersionFile(AcidTables.java:91)
ibis-trino-1  |         at io.trino.plugin.hive.HiveMetadata.createOrcAcidVersionFile(HiveMetadata.java:2326)
ibis-trino-1  |         ... 19 more
ibis-trino-1  | Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Object name contains unsupported characters. (Service: Amazon S3; Status Code: 400; Error Code: XMinioInvalidObjectName; Request ID: 178E502539DD4537; S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8; Proxy: null)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1879)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1418)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1387)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1157)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:814)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)
ibis-trino-1  |         at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5520)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5467)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client.access$300(AmazonS3Client.java:422)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client$PutObjectStrategy.invokeServiceCall(AmazonS3Client.java:6601)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client.uploadObject(AmazonS3Client.java:1891)
ibis-trino-1  |         at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1851)
ibis-trino-1  |         at io.trino.hdfs.s3.TrinoS3FileSystem$TrinoS3StreamingOutputStream.flushBuffer(TrinoS3FileSystem.java:1909)
ibis-trino-1  |         ... 27 more
ibis-trino-1  | 2023-10-15T15:03:50.419Z        ERROR   transaction-finishing-1 io.trino.metadata.CatalogMetadata       Connector threw exception on abort
ibis-trino-1  | java.lang.IllegalStateException: Tried to rollback buffered metastore operations after transaction has been committed/aborted
ibis-trino-1  |         at io.trino.plugin.hive.metastore.SemiTransactionalHiveMetastore.rollback(SemiTransactionalHiveMetastore.java:1314)
ibis-trino-1  |         at io.trino.plugin.hive.HiveMetadata.rollback(HiveMetadata.java:3761)
ibis-trino-1  |         at io.trino.plugin.hive.HiveTransactionManager.lambda$rollback$1(HiveTransactionManager.java:69)
ibis-trino-1  |         at java.base/java.util.Optional.ifPresent(Optional.java:178)
ibis-trino-1  |         at io.trino.plugin.hive.HiveTransactionManager.rollback(HiveTransactionManager.java:67)
ibis-trino-1  |         at io.trino.plugin.hive.HiveConnector.rollback(HiveConnector.java:226)
ibis-trino-1  |         at io.trino.metadata.CatalogTransaction.abort(CatalogTransaction.java:92)
ibis-trino-1  |         at io.trino.metadata.CatalogMetadata.safeAbort(CatalogMetadata.java:167)
ibis-trino-1  |         at io.trino.metadata.CatalogMetadata.abort(CatalogMetadata.java:161)
ibis-trino-1  |         at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
ibis-trino-1  |         at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
ibis-trino-1  |         at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:75)
ibis-trino-1  |         at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
ibis-trino-1  |         at io.airlift.concurrent.BoundedExecutor.drainQueue(BoundedExecutor.java:79)
ibis-trino-1  |         at io.trino.$gen.Trino_429____20231015_145128_2.run(Unknown Source)
ibis-trino-1  |         at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
ibis-trino-1  |         at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
ibis-trino-1  |         at java.base/java.lang.Thread.run(Thread.java:833)
```
</details>

From those logs I noticed

```
Caused by: com.amazonaws.services.s3.model.AmazonS3Exception: Object name contains unsupported characters. (Service: Amazon S3; Status Code: 400; Error Code: XMinioInvalidObjectName; Request ID: 178E502539DD4537; S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8; Proxy: null)
```

and that double slash looked a bit suspicious so I poked around the internet a bit.

I am using minio for storage (this is a development setup for testing [Ibis](https://github.com/ibis-project/ibis)), and it turns out that minio [doesn't support duplicate slashes in object paths](https://github.com/minio/minio/issues/5874).

I poked around the Trino codebase a bit, and noticed these lines:

https://github.com/trinodb/trino/blob/91ee252b8eba1864e35661d57a854be06b98fe27/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java#L2132-L2135

which seems to assume that transactional tables are partitioned, but since the one I created isn't, the `update.getName()` bit returns the empty string and the format string ends up being `%s//%s` resulting in the double slash that Minio doesn't accept.

Indeed, trying this on a partitioned table, I can insert rows into it without a problem:

```
trino:default> CREATE TABLE t (x INT, y INT) WITH (transactional = true, partitioned_by = array['y']);
CREATE TABLE
trino:default> INSERT INTO t VALUES (1, 2), (1, 3);
INSERT: 2 rows

Query 20231015_150747_01334_4bjpf, FINISHED, 1 node
Splits: 36 total, 36 done (100.00%)
0.17 [0 rows, 0B] [0 rows/s, 0B/s]
```

--- a/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java
+++ b/plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java
@@ -2131,7 +2131,11 @@ public class HiveMetadata
         if (isFullAcidTable(table.getParameters())) {
             for (PartitionUpdate update : partitionUpdates) {
                 String deltaSubdir = deltaSubdir(handle.getTransaction().getWriteId(), 0);
-                String directory = "%s/%s/%s".formatted(table.getStorage().getLocation(), update.getName(), deltaSubdir);
+                String directory = table.getStorage().getLocation();
+                if (!update.getName().isEmpty()) {
+                    directory += "/" + update.getName();
+                }
+                directory += "/" + deltaSubdir;
                 createOrcAcidVersionFile(session.getIdentity(), directory);
             }
         }

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestWriteToHiveTransactionalTableInTrino.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/hive/TestWriteToHiveTransactionalTableInTrino.java
@@ -47,6 +47,16 @@ public class TestWriteToHiveTransactionalTableInTrino
     }
 
     @Test(groups = {HMS_ONLY, PROFILE_SPECIFIC_TESTS})
+    public void testInsertIntoNonPartitionedTable()
+    {
+        String tableName = "non_partitioned_transactional_insert";
+        onTrino().executeQuery(format("CREATE TABLE %s (column1 INT, column2 INT) WITH (transactional = true)", tableName));
+        onTrino().executeQuery(format("INSERT INTO %s VALUES (11, 12), (111, 121)", tableName));
+        assertThat(onTrino().executeQuery(format("SELECT * FROM %s", tableName))).containsOnly(row(11, 12), row(111, 121));
+        onTrino().executeQuery(format("DROP TABLE %s", tableName));
+    }
+
+    @Test(groups = {HMS_ONLY, PROFILE_SPECIFIC_TESTS})
     public void testUpdateOnUnpartitionedTable()
     {
         String tableName = "unpartitioned_transactional_update";

